### ISSUE TITLE
Discovery failure, but snmpwalk is working - Sophos UTM

### ISSUE THREAD

**Issue by ikus060** [netdisco/netdisco#312]
Overall, my discovery process is working as expected. I'm able to discover by Switchs and APs, but the router cannot be discovered by Netdisco. Meanfile, I'm pretty sure it's not a SNMO community issue since snmpwalk is working as expected.

## Expected Behavior
I'm expecting the discovery of the my router to work.
`~/bin/netdisco-do discover -D -d 192.168.14.1`

## Current Behavior
The discovery is failing as follow:
```
[2002] 2017-06-12 19:21:04  info App::Netdisco version 2.035006 loaded.
[2002] 2017-06-12 19:21:04  info discover: started at Mon Jun 12 19:21:04 2017
[2002] 2017-06-12 19:21:04 debug [192.168.14.1] try_connect with ver: 2, class: SNMP::Info, comm: <hidden>
[2002] 2017-06-12 19:21:04 debug [192.168.14.1] try_connect with ver: 1, class: SNMP::Info, comm: <hidden>
[2002] 2017-06-12 19:21:04 debug [192.168.14.1] try_connect with ver: 2, class: SNMP::Info, comm: <hidden>
[2002] 2017-06-12 19:21:04 debug [192.168.14.1] try_connect with ver: 1, class: SNMP::Info, comm: <hidden>
[2002] 2017-06-12 19:21:04 debug [192.168.14.1] try_connect with ver: 2, class: SNMP::Info, comm: <hidden>
[2002] 2017-06-12 19:21:04 debug [192.168.14.1] try_connect with ver: 1, class: SNMP::Info, comm: <hidden>
[2002] 2017-06-12 19:21:04  info discover: finished at Mon Jun 12 19:21:04 2017
[2002] 2017-06-12 19:21:04  info discover: status error: discover failed: could not SNMP connect to 192.168.14.1
```

Take note, an snmpwalk from the same server is working.
```
snmpwalk -v 2c -c public 192.168.14.1
iso.3.6.1.2.1.1.1.0 = ""
iso.3.6.1.2.1.1.2.0 = OID: iso.3.6.1.4.1.21067.2
iso.3.6.1.2.1.1.3.0 = Timeticks: (18385390) 2 days, 3:04:13.90
```

I'm suspecting the model of device to be an issue. It's a Sophos UTM. Maybe the SNMP is mis implemented ? Is Netdisco looking for a specific oid ?

**Comment by ollyg** [comment]
Hi,

You can check the SNMP community used is correct by running the job with SHOW_COMMUNITY=1 in the environment. You can also enable SNMP debugging:

`SHOW_COMMUNITY=1 ~netdisco/bin/netdisco-do discover -d 192.168.14.1 -D -I`

Add another `-I` switch for more SNMP debugging.

Try that and see how it goes.

**Comment by ikus060** [comment]
Hello ollyg,

Thanks alot for you help. I've tried various thing. As follow:
```
SHOW_COMMUNITY=1 ~netdisco/bin/netdisco-do discover -d 192.168.14.1 -D -I
[3027] 2017-06-12 22:06:32  info App::Netdisco version 2.035006 loaded.
[3027] 2017-06-12 22:06:33  info discover: started at Mon Jun 12 22:06:33 2017
[3027] 2017-06-12 22:06:33 debug [192.168.14.1] try_connect with ver: 2, class: SNMP::Info, comm: public
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
[3027] 2017-06-12 22:06:33 debug [192.168.14.1] try_connect with ver: 1, class: SNMP::Info, comm: public
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
[3027] 2017-06-12 22:06:33 debug [192.168.14.1] try_connect with ver: 2, class: SNMP::Info, comm: M86YcsUhkN9rMgJH
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
[3027] 2017-06-12 22:06:33 debug [192.168.14.1] try_connect with ver: 1, class: SNMP::Info, comm: M86YcsUhkN9rMgJH
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
[3027] 2017-06-12 22:06:33 debug [192.168.14.1] try_connect with ver: 2, class: SNMP::Info, comm: public
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
[3027] 2017-06-12 22:06:33 debug [192.168.14.1] try_connect with ver: 1, class: SNMP::Info, comm: public
SNMP::Info::_global uptime : DISMAN-EVENT-MIB::sysUpTimeInstance : .1.3.6.1.2.1.1.3.0
SNMP::Info::_global layers : SNMPv2-MIB::sysServices.0 : .1.3.6.1.2.1.1.7.0
SNMP::Info::_global description : SNMPv2-MIB::sysDescr.0 : .1.3.6.1.2.1.1.1.0
[3027] 2017-06-12 22:06:33  info discover: finished at Mon Jun 12 22:06:33 2017
[3027] 2017-06-12 22:06:33  info discover: status error: discover failed: could not SNMP connect to 192.168.14.1
```
Using SNMP Walk:
```
snmpwalk -v 1 -c public 192.168.14.1 .1.3.6.1.2.1.1.3.0
iso.3.6.1.2.1.1.3.0 = Timeticks: (19355660) 2 days, 5:45:56.60

snmpwalk -v 1 -c public 192.168.14.1 .1.3.6.1.2.1.1.7.0
iso.3.6.1.2.1.1.7.0 = INTEGER: 0

snmpwalk -v 1 -c public 192.168.14.1 .1.3.6.1.2.1.1.1.0
iso.3.6.1.2.1.1.1.0 = ""
```

Do you think the discovery is failing because my description is empty ?

**Comment by ikus060** [comment]
That is confirm, the discovery is failing if the description is empty. I change the value in Sophos WebUI to something random and the discovery pass.

**Comment by ollyg** [comment]
Many thanks! This is an interesting case.

The SNMP::Info Perl module requests several OIDs from the device to determine the "device type". I guess it dies if it cannot get a value for the sysDescr. I'm going to leave this ticket open and ask another developer whether we consider this a bug, or not (in the IRC channel if you want to join).

Thanks again,
Oliver.

**Comment by JeroenvIS** [comment]
IMHO the Sophos UTM is too loose in its SNMP implementation; I'd rather treat this as a Sophos bug than as a bug in SNMP::Info.

If I understand correctly, the Sophos UTM had an empty sysDescr.0 value by default. The description of that object is:

> SNMPv2-MIB::sysDescr
> sysDescr OBJECT-TYPE
>   -- FROM	SNMPv2-MIB, RFC1213-MIB
>   -- TEXTUAL CONVENTION DisplayString
>   SYNTAX	OCTET STRING (0..255) 
>   MAX-ACCESS	read-only
>   STATUS	mandatory
>   DESCRIPTION	"A textual description of the entity.  This value
>             should include the full name and version
>             identification of the system's hardware type,
>             software operating-system, and networking
>             software.  It is mandatory that this only contain
>             printable ASCII characters."
> ::= { iso(1) org(3) dod(6) internet(1) mgmt(2) mib-2(1) system(1) 1 }

...so it 'should' by default include a a certain level of detail about the hardware and software that the SNMP agent is running on. The object is read-only; virtually all vendors indeed use a format that describes the platform and OS version on it, and don't allow changing it to any other user-settable value; not through SNMP, nor through any other means of configuration.

Of course we could consider working around this implementation, but given that the device _also_ doesn't report a useful value in sysServices, I'd advise against that; we'd be stacking workaround upon workaround.

Basically the Sophos says by default "I'm a device that doesn't want to identify myself in a commonly accepted way, and I don't provide any network services at any layer of the OSI model".

@ollyg, what's your position on this?

**Comment by JeroenvIS** [comment]
@ikus060: thanks for the detailed report btw!

**Comment by ikus060** [comment]
I agree with you, the SNMP implementation in Sophos UTM is very slack. As a workarround, it's possible to define a value in the sysDescr using the Sophos web interface.

At least, now their is some trace of the issue in github.

**Comment by JeroenvIS** [comment]
Great! Let's close this one then.

**Comment by ollyg** [comment]
Thanks! :-) FWIW, I agree.
